### PR TITLE
Add support for custom ONNX GroupNorm operator

### PR DESCRIPTION
**Pull request by mateusztabaka**

### CODE REVIEW

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    namespace onnx_import
+    {
+        namespace op
+        {
+            namespace detail
+            {
+                namespace
+                {
+                    std::shared_ptr<ngraph::Node>
+                        create_group_norm_shape(const Output<ngraph::Node>& data, size_t num_groups)
```
Comment (postrational) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:36]: Please add a comment explaining what this method should return.

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                                       num_groups_const,
+                                       std::make_shared<default_opset::Divide>(splits[1],
+                                                                               num_groups_const),
+                                       splits[2].get_node_shared_ptr(),
+                                       splits[3].get_node_shared_ptr()},
+                            0);
+                    }
+
+                    std::shared_ptr<ngraph::Node>
+                        create_scale_bias_shape(const Output<ngraph::Node>& node,
```
Comment (postrational) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:66]: Please add a comment here as well.

**Review by postrational — CHANGES_REQUESTED** [review]
Please also add `GroupNorm` to `onnx.cpp` to assist with models which don't specify the custom domain for the op.

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                {
+                    // This function creates a shape to which we need to reshape the input
+                    // before normalization.
+                    // If data shape is [N, C, H, W], then this function returns [N, num_groups, C // num_groups, H, W]
+                    std::shared_ptr<ngraph::Node>
+                        create_group_norm_shape(const Output<ngraph::Node>& data, size_t num_groups)
+                    {
+                        const auto& pshape = data.get_partial_shape();
+                        size_t rank_size = pshape.rank().get_length();
+                        NGRAPH_CHECK(rank_size == 4, "Only 4-D input tensors supported");
```
Comment (tsocha) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:44]: Why are you support only 4D tensors? It's not so hard to make it work for the larger ones.
You can use the solution from the `create_scale_bias_shape` function.

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                    }
+                }
+            } // detail
+
+            namespace set_1
+            {
+                OutputVector group_norm(const Node& node)
+                {
+                    auto inputs = node.get_ng_inputs();
+                    NGRAPH_CHECK(inputs.size() == 3, "Invalid number of inputs");
```
Comment (tsocha) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:134]: Please add the number of provided and expected number of inputs in the description.

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                        bias = std::make_shared<default_opset::Reshape>(
+                            bias, detail::create_scale_bias_shape(bias, data_rank_size), true);
+                    }
+
+                    std::shared_ptr<ngraph::Node> result =
+                        std::make_shared<default_opset::Divide>(diff, sqrt);
+                    result =
+                        std::make_shared<default_opset::Reshape>(result, data_shape_node, true);
+                    result = std::make_shared<default_opset::Multiply>(scale, result);
+                    result = std::make_shared<default_opset::Add>(result, bias);
```
Comment (tsocha) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:193]: Could you check if this helper can be used here?
https://github.com/openvinotoolkit/openvino/blob/master/ngraph/frontend/onnx_import/src/op/conv.cpp#L80-L99

**Review by tsocha — CHANGES_REQUESTED** [review]

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 6]
```diff
@@ -41,13 +42,17 @@ namespace ngraph
...
 
                         if (pshape.is_static())
                         {
                             const auto& shape = pshape.to_shape();
                             std::vector<size_t> new_shape{
                                 shape[0], num_groups, shape[1] / num_groups, shape[2], shape[3]};
+                            for (size_t i = 4; i < rank_size; i++)
+                            {
+                                new_shape.push_back(shape[i]);
+                            }
```
Comment (tsocha) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp]:
```suggestion
                        NGRAPH_CHECK(rank_size >= 2, "2-D and above tensors supported only");

                        if (pshape.is_static())
                        {
                            const auto& shape = pshape.to_shape();
                            std::vector<size_t> new_shape{
                                shape[0], num_groups, shape[1] / num_groups};
                            for (size_t i = 2; i < rank_size; i++)
                            {
                                new_shape.push_back(shape[i]);
                            }
```

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 7]
```diff
@@ -56,72 +61,17 @@ namespace ngraph
...
-                        auto one = default_opset::Constant::create(element::i64, Shape{1}, {1});
-                        NodeVector dims;
-                        if (rank_size == 1)
+                        NodeVector new_shape{
+                            splits[0].get_node_shared_ptr(),
+                            num_groups_const,
+                            std::make_shared<default_opset::Divide>(splits[1], num_groups_const),
+                            splits[2].get_node_shared_ptr(),
+                            splits[3].get_node_shared_ptr()};
+                        for (size_t i = 4; i < rank_size; i++)
```
Comment (tsocha) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:70]:
```suggestion
                        for (size_t i = 2; i < rank_size; i++)
```

**Review by tsocha — CHANGES_REQUESTED** [review]

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp (hunk 6 above)
Comment (mateusztabaka) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp]: why do we need 2-d or 3-d tensors in group norm?

Comment (tsocha) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp]: If it's not prohibited by the spec(if there is any) it should be allowed.
3D tensors can be useful for text data or sound wave for example.
2D tensors are not useful at all in this context and can be prohibited in the line no. 55.
There is no need to write indexes 2 and 3 by hand if this can be handled by this for loop if you are afraid of performance of 2 more push_backs then you can reserve the memory for it.

Comment (mateusztabaka) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp]: > 2D tensors are not useful at all in this context and can be prohibited in the line no. 55.
so the line should be rather look like:
NGRAPH_CHECK(rank_size >= 3, "3-D and above tensors supported only");

**Review by tsocha — APPROVED** [review]

File: ngraph/frontend/onnx_import/include/onnx_import/utils/reshape.hpp [hunk 8]
```diff
@@ -61,6 +61,20 @@ namespace ngraph
...
+            /// \note       Usually scale or bias to BatchNorm or related operators are
+            ///             nodes of shape {C}. This function will reshape the input node
+            ///             with a shape of {C} into a node with Shape{1, C, 1, 1, ..}.
+            ///
+            /// \param[in]  scale_or_bias   Node to reshape.
+            /// \param[in]  expected_rank   Expected rank size
+            ///
+            /// \return     Original node or a node representing a reshape of the original.
+            ///
+            Output<ngraph::Node> reshape_scale_or_bias(const Output<ngraph::Node>& scale_or_bias,
```
Comment (postrational) [inline ngraph/frontend/onnx_import/include/onnx_import/utils/reshape.hpp:75]: Let's give this function a name related to what it does, not how it's used, please.

**Review by postrational — CHANGES_REQUESTED** [review]

**Review by postrational — APPROVED** [review]

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                {
+                    // This function creates a shape to which we need to reshape the input
+                    // before normalization.
+                    // If data shape is [N,C,H,W], the function returns
+                    // [N, num_groups, C // num_groups, H, W]
+                    std::shared_ptr<ngraph::Node>
+                        create_group_norm_shape(const Output<ngraph::Node>& data, size_t num_groups)
+                    {
+                        const auto& pshape = data.get_partial_shape();
+                        size_t rank_size = pshape.rank().get_length();
```
Comment (mbencer) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:44]: Please add assert which check if rank is static

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+                            diff, default_opset::Constant::create(element::f32, Shape{}, {2})),
+                        reduction_axes,
+                        true);
+
+                    const std::shared_ptr<ngraph::Node> eps_node =
+                        std::make_shared<default_opset::Constant>(element::f32, Shape{}, eps);
+                    const auto sqrt = std::make_shared<default_opset::Sqrt>(
+                        std::make_shared<default_opset::Add>(variance, eps_node));
+
+                    auto data_rank_size = data.get_partial_shape().rank().get_length();
```
Comment (mbencer) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:124]: As above

File: ngraph/frontend/onnx_import/src/utils/reshape.cpp [hunk 11]
```diff
@@ -114,6 +114,23 @@ namespace ngraph
                 return builder::opset1::reshape(node, Shape{});
             }
 
+            Output<ngraph::Node>
+                reshape_channel_shaped_node_to_nchw(const Output<ngraph::Node>& node,
+                                                    size_t expected_rank)
+            {
+                size_t node_rank = node.get_partial_shape().rank().get_length();
```
Comment (mbencer) [inline ngraph/frontend/onnx_import/src/utils/reshape.cpp:121]: as above ;)

**Review by mbencer — APPROVED** [review]

File: ngraph/frontend/onnx_import/src/onnx.cpp [hunk 12]
```diff
@@ -74,7 +74,7 @@ namespace ngraph
             } // namespace error
 
             static const std::vector<std::string> legacy_ops_to_fixup = {
-                "FakeQuantize", "DetectionOutput", "Normalize", "PriorBox"};
+                "DetectionOutput", "FakeQuantize", "GroupNorm", "Normalize", "PriorBox"};
```
Comment (slyalin) [inline ngraph/frontend/onnx_import/src/onnx.cpp:77]: @postrational, finally, have you found a reliable description for all those operations?

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+                                 "Invalid number of inputs. Expected 3, actual " +
+                                     std::to_string(inputs.size()));
+
+                    auto data = inputs[0];
+                    auto scale = inputs[1];
+                    auto bias = inputs[2];
+
+                    size_t num_groups =
+                        static_cast<size_t>(node.get_attribute_value<int64_t>("num_groups"));
+                    float eps = node.get_attribute_value<float>("eps", 1e-5);
```
Comment (slyalin) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:93]: In the code of the MO extractor for this op (https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/extensions/front/onnx/group_norm_ext.py#L45), the default value for eps is 1e-6. Why is it 1e-5 here?

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+                    auto data_pshape = data.get_partial_shape();
+                    std::shared_ptr<ngraph::Node> data_shape_node;
+                    if (data_pshape.is_static())
```
Comment (slyalin) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:97]: I don't think it is correct. Even if shape is static we may want to reshape a model later before loading an nGraph function to a plugin. We have to keep shape sub-expressions as many as possible to maintain this capability.

File: ngraph/test/models/onnx/group_norm.prototxt [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+            dim_value: 2
+          }
+        }
+      }
+    }
+  }
+}
+opset_import {
+  version: 1
+}
```
Comment (slyalin) [inline ngraph/test/models/onnx/group_norm.prototxt:108]: Can we pass this file to MO and compare results from MO and ONNX importer for it? I wasn't able to pass this file to MO, text pb is not supported. @mateusztabaka, could you suggest a tool to make it possible?

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+
+                    std::shared_ptr<ngraph::Node> result =
+                        std::make_shared<default_opset::Divide>(diff, sqrt);
+                    result =
+                        std::make_shared<default_opset::Reshape>(result, data_shape_node, true);
+                    result = std::make_shared<default_opset::Multiply>(
+                        reshape::reshape_channel_shaped_node_to_nchw(scale, data_rank_size),
+                        result);
+                    result = std::make_shared<default_opset::Add>(
+                        result, reshape::reshape_channel_shaped_node_to_nchw(bias, data_rank_size));
```
Comment (slyalin) [inline ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:137]: Why is the resulting sub-graph generated different from what MO does in this transformation: https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/extensions/middle/GroupNorm.py. Does MO extra conversion of MVN later on in the transformation pipeline? @mateusztabaka, have you checked that?

**Review by slyalin — COMMENTED** [review]
@mateusztabaka, @postrational please consider reverting this PR or respond with justifications to my comments.

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp (hunk 13 above)
Comment (mateusztabaka) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:93]: I was suggested by the paper https://arxiv.org/pdf/1803.08494.pdf and pytorch definition https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html#torch.nn.GroupNorm - we have 1e-5 there. 
I can adjust those two. Which value should I pick?

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp (hunk 14 above)
Comment (mateusztabaka) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:97]: Take a look at group norm paper https://arxiv.org/pdf/1803.08494.pdf. By its definition, you have to apply reshape to the input.

File: ngraph/test/models/onnx/group_norm.prototxt (hunk 15 above)
Comment (mateusztabaka) [reply ngraph/test/models/onnx/group_norm.prototxt:108]: you can use this tool https://github.com/openvinotoolkit/openvino/blob/master/ngraph/test/models/onnx/onnx_prototxt_converter.py

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp (hunk 16 above)
Comment (mateusztabaka) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:137]: It's different, because the one from MO is not correct. If you take a look at https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/normalization/MVN_1.md and the paper https://arxiv.org/pdf/1803.08494.pdf, we can't really use MVN here. There is a difference in denominator: MVN uses sqrt(x) + eps, while GroupNorm requires sqrt(x + eps). 
MO implementation works on CPU only by accident - there is a bug in MKLDNN plugin: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/mkldnn_plugin/nodes/mkldnn_mvn_node.cpp#L815.
I could fix that bug, but there are few transformations in MO that rely on it right now.

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp (hunk 14 above)
Comment (mateusztabaka) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:97]: That's also what MO does.

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp (hunk 13 above)
Comment (slyalin) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:93]: I don't know why MO uses that value. @lazarevevgeny probably knows. At least we have to use the same default value in all components.

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp (hunk 14 above)
Comment (slyalin) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:97]: I'm not arguing that such reshape is not needed. I think that we always should follow the approach where ShapeOf is used and shouldn't use shape from partial_shape of a node even if it is statically known during the conversion. MO uses it at https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/extensions/middle/GroupNorm.py#L60. This code looks to be used unconditionally.

File: ngraph/test/models/onnx/group_norm.prototxt (hunk 15 above)
Comment (slyalin) [reply ngraph/test/models/onnx/group_norm.prototxt:108]: @evgenytalanin-intel, please consider using this tool to sync coverage between the importer and MO.

Comment (slyalin) [reply ngraph/test/models/onnx/group_norm.prototxt:108]: @mateusztabaka, thanks, works perfectly for me. But still there is a question about coverage and comparison to MO. Can @evgenytalanin-intel help us?

File: ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp (hunk 16 above)
Comment (slyalin) [reply ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp:137]: @mateusztabaka by "fixing the bug" do you mean fixing mkldnn bug or change in GroupNorm conversion? In this case we should also change InstanceNormalization from ONNX, because it should suffer from the same bug. Can we make sure that this is really critical difference in real-life models?
